src/agent/platform.rs

```diff
@@ -136,172 +141,133 @@ impl Client {
-        Ok(Self{token})
-    }
+struct EnrolledToken {
+    refresh_token: String,
+    session_token: String,
+    expiration: SystemTime,
+}
 
-    fn save(&self) -> Result<()> {
-        let token_file = PathBuf::from(TOKEN_FILE);
-        let dir = token_file.parent().unwrap();
-        std::fs::create_dir_all(dir)?;
-        Ok(std::fs::write(TOKEN_FILE, &self.token)?)
-    }
+async fn enroll(channel: Channel, deploy_token: String, hostname: String, machine_id: String) -> Result<EnrolledToken> {
+    let mut token_svc = TokenServiceClient::new(channel);
+
+    let req = pb::EnrollAgentRequest {
+        deployment_token: deploy_token,
+        hostname,
+        agent_version: VERSION.to_string(),
+        machine_id,
+    };
+
+    let resp = token_svc.enroll_agent(req).await
+        .map_err(|e| anyhow!("{}", e.message()))?
+        .into_inner();
+
+    // ensure the token is ascii
+    _ = AsciiMetadataValue::try_from(&resp.session_token)
+        .map_err(|_| anyhow!("session token is not ASCII"))?;
+
+    Ok(EnrolledToken{
+        refresh_token: resp.refresh_token,
+        session_token: resp.session_token,
+        expiration: get_expiration(resp.session_token_expiration),
+    })
+}
 
-    fn get(&self) -> String {
-        self.token.clone()
+async fn enroll_loop(channel: Channel, deploy_token: String, hostname: String, machine_id: String) -> EnrolledToken {
+    loop {
+        match enroll(channel.clone(), deploy_token.clone(), hostname.clone(), machine_id.clone()).await {
+            Ok(tok) => return tok,
+            Err(err) => {
+                error!("Agent enrollment failed: {err}");
+                tokio::time::sleep(RETRY_INTERVAL).await;
+            }
+        }
     }
 }
 
-struct SessionKeeper {
-    refresh_token: RefreshToken,
-    auth_token: AuthToken,
-    expiration: SystemTime,
+async fn refresh_loop(
     channel: Channel,
-}
+    refresh_token: String,
+    auth_token: AuthToken,
+    mut expiration: SystemTime,
+) -> Result<()> {
+    let mut token_svc = TokenServiceClient::with_interceptor(channel, auth_token.clone());
 
-impl SessionKeeper {
-    async fn new(channel: Channel, deploy_token: String, hostname: String, machine_id: String) -> Result<Self> {
-        let mut token_svc = TokenServiceClient::new(channel.clone());
+    loop {
+        let mut interval = expiration
+            .duration_since(SystemTime::now())
+            .unwrap_or_else(|_| {
+                error!("Session expiration is in the past");
+                DEFAULT_EXPIRATION
+            });
 
-        let (refresh_token, session_token, expiration) = match RefreshToken::load() {
-            Ok(refresh_token) => {
-                let req = pb::GetSessionTokenRequest{
-                    refresh_token: refresh_token.get(),
-                    agent_version: VERSION.to_string(),
-                };
+        interval = interval.checked_sub(EXPIRATION_SLACK).unwrap_or(interval);
```

Comment:
If interval is less than EXPIRATION_SLACK, shouldn't the interval be set to now?

Reply:
Seems bad to refresh every time. If the expiration is in 30 mins, I figured to just go ahead and do it 30 mins. Perhaps a better strategy is to use a percentage based one.

Reply:
I was thinking of the case where expiration is in ten minutes but the slack is thirty minutes, which would lead to the token being invalid for the last twenty minutes before it's refreshed. But I guess that's a good rate-limiting mechanism, since the expiration should never _repeatedly_ be less than the slack.

Reply:
Oh sure, but in this case if the expiration is in 10 minutes but the slack is 30, it will renew in 10 minutes.

Reply:
Ohhh, wow I was misreading that. Yep, that seems fine.